docs/current/guides/899944-aws-cdk-ecs.md

```diff
@@ -0,0 +1,114 @@
+---
+slug: /899944/aws-cdk-ecs
+displayed_sidebar: "current"
+category: "guides"
+tags: ["go", "aws", "cdk", "ecs", "fargate"]
+authors: ["Sam Alba"]
+date: "2023-03-02"
+---
+
+# Use Dagger with the AWS CDK (Cloud Development Kit)
+
+:::note
+[Watch a live demo](https://youtu.be/ESZKu8VWSGA) of this tutorial in the Dagger Community Call (23 Feb 2023). For more demos, [join the next Dagger Community Call](https://dagger.io/events).
+:::
+
+## Introduction
+
+[The AWS CDK](https://docs.aws.amazon.com/cdk/v2/guide/home.html) is a framework that enables developers to use their programming language of choice to describe Infrastructure resources on AWS. However there are several things that the CDK will not support, for example: building your application, running your tests, manage your container images, etc... This is where Dagger comes in handy.
```

Comment:
Can you please give an example of application that CDK can not build? In my experience building docker image, or bundling a TypeScript app into a Lambda, or Rust binary also in Lambda are all relatively simple tasks in CDK (can bring up examples in the thread).

Also, there is an official pipeline lib by AWS for CDK deployments. In general — are there advantages of mixing Dagger with CDK?

Reply:
Hey Artem, nice to hear from you, it's been a while! :-)

Few thoughts about your questions:

> Can you please give an example of application that CDK can not build?

In my experience, the CDK does not implement a complete build system like Dagger provides (DAG pipeline, caching, parallelization), ideally this is how you want to run your tests, generating assets, etc... Although the CDK provides some nice helpers: `lambda.DockerImageCode.fromImageAsset(<local path>)`.

> Also, there is [an official pipeline lib by AWS for CDK deployments](https://docs.aws.amazon.com/cdk/v2/guide/cdk_pipeline.html). In general — are there advantages of mixing Dagger with CDK?

Sure, this is the service AWS Code Pipeline being exposed in the CDK, which will produce a CloudFormation template. And this pipeline won't run outside of AWS (eg. on your local machine).

Reply:
> nice to hear from you, it's been a while! :-)

It's been a while, yeah! Quite happy to talk again to be honest.

Thank you for the detailed explanation. I still have some questions regarding the Dagger+CDK integration. Pardon for the "long read", tried to be as less Hamingway-ish as possible😬

**Assets and builds**
CDK allows building for quite a lot of targets in a declarative way. A few of them:
* Dockerfile > container (ECS/EKS/Lambda/...)
* TypeScript > JS asset (Lambda)
* Custom runtime (Rust/Zig/...) > binary asset (Lambda).

Basically you just point a construct at a dir with your code, and at the stage of CFN synth it builds your code, bundles it in assets and then uploads to S3 (or ECR for example) for you. So it does [this](https://github.com/dagger/examples/blob/0126559860d11888b142dda30f02f72ebeeec317/go/aws-cdk/main.go#L40-L51) part out of the box.

It also gives you a freedom of building environment — you can either build on a host machine, or build your assets in a docker container, collect outputs from there and then upload them to AWS.

**Caching and parallelization**
Built assets are cached. You are also free to implement a custom hashing algorithm. Synthesised templates are not cached though as far as I know, but synth normally takes seconds.

Parallelization is trickier. While stacks are deployed in parallel (if no dependencies), builds of their lambda/containers happen sequentially from my experience.

> ❓Is this something Dagger is supposed to help? If yes, how can I start bundling/building lambdas in parallel with minimal changes to the existing CDK codebase?

**Tests**
CDK gives a developer tooling for [infrastructure testing](https://docs.aws.amazon.com/cdk/v2/guide/testing.html), which essentially is validation of CFN templates.

Also official lambda constructs [allow for running commands](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda_nodejs.ICommandHooks.html) at the different stages of bundling. I admit I can miss something, but not quite sure how Dagger complements CDK in the sense of testing.

Reply:
What you describe is a complete alternative with the CDK doing more than infra management, and without dagger. I'll return the ask to see examples of what you describe in practice.

Most users, including myself, think of the CDK as a CloudFormation wrapper. CFN is a great reconciliation engine for your aws stack, but it's a pain to use, and the CDK is the great improvement.

I'll refine the wording because I don't want to bake personal opinions into a guide that shows a way to solve a problem. You could also say, all of this is possible with 100% GitHub actions.

But from my experience, and it's likely biased, I prefer to let dagger handle the CI side of things.

Reply:
Got it, thank you!